### PR TITLE
IGNITE-23591 Fix entry and checkpoint locks orders

### PR DESCRIPTION
**Pull request by timoninmaxim**
Thank you for submitting the pull request to the Apache Ignite.

In order to streamline the review of the contribution 
we ask you to ensure the following steps have been taken:

### The Contribution Checklist
- [ ] There is a single JIRA ticket related to the pull request. 
- [ ] The web-link to the pull request is attached to the JIRA ticket.
- [ ] The JIRA ticket has the _Patch Available_ state.
- [ ] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [ ] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/ConcurrentCheckpointAndUpdateTtlTest.java [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+        }, 1, "update");
+
+        IgniteInternalFuture<?> touchFut = runMultiThreadedAsync(() -> {
+            while (!stop.get()) {
+                if (touchWithScanQuery)
+                    cln.cache(DEFAULT_CACHE_NAME).query(new ScanQuery<>()).getAll();
+                else
+                    cln.cache(DEFAULT_CACHE_NAME).get(KEY);
+            }
+
```
Comment (alex-plekhanov) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/ConcurrentCheckpointAndUpdateTtlTest.java:142]: Redundant NL

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/ConcurrentCheckpointAndUpdateTtlTest.java [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+        try (IgniteDataStreamer<Integer, Integer> stream = cln.dataStreamer(DEFAULT_CACHE_NAME)) {
+            // Skip annoying warnings.
+            stream.allowOverwrite(true);
+
+            Map<Integer, Integer> map = new HashMap<>();
+
+            map.put(KEY, rnd.nextInt());
+
+            stream.addData(map);
+            stream.flush();
```
Comment (alex-plekhanov) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/ConcurrentCheckpointAndUpdateTtlTest.java:174]: when using try-with-resource flush is redundant

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/ConcurrentCheckpointAndUpdateTtlTest.java [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+        stop.set(true);
+
+        GridTestUtils.waitForAllFutures(updateFut, touchFut, cpFut);
+    }
+
+    /** */
+    private void stream(Ignite cln) {
+        try (IgniteDataStreamer<Integer, Integer> stream = cln.dataStreamer(DEFAULT_CACHE_NAME)) {
+            // Skip annoying warnings.
+            stream.allowOverwrite(true);
```
Comment (alex-plekhanov) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/ConcurrentCheckpointAndUpdateTtlTest.java:167]: With allowOverwrite=true we test the same code branch (cache put method). But with allowOverwrite=false another code branch can be checked (ISOLATED_UPDATER), so, I propose to use default value (false) here.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/ConcurrentCheckpointAndUpdateTtlTest.java [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+
+            Map<Integer, Integer> map = new HashMap<>();
+
+            map.put(KEY, rnd.nextInt());
+
+            stream.addData(map);
```
Comment (alex-plekhanov) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/ConcurrentCheckpointAndUpdateTtlTest.java:173]: Why not `addData(K key, V val)`? In this case we don't need map creation.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/ConcurrentCheckpointAndUpdateTtlTest.java [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+                try {
+                    forceCheckpoint(F.asList(grid(0), grid(1)));
+                }
+                catch (Exception e) {
+                    // No-op.
+                }
+            }
+        }, 1, "checkpoint");
+
+        Thread.sleep(30_000);
```
Comment (alex-plekhanov) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/ConcurrentCheckpointAndUpdateTtlTest.java:156]: 8 tests * 30 secs, 4 minutes total time. Perhaps too much, WDYT?

Comment (timoninmaxim) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/cache/ConcurrentCheckpointAndUpdateTtlTest.java]: On my machine it requires 10-20 secs to reproduce. Replace it with counter to 1000 checkpoint iterations.

**Review by alex-plekhanov — APPROVED** [review]

**Pull request merged**